File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> unique = new HashSet<Character>();
        int size = 0;
        int start = 0;
        for (int end = 0; end < s.length(); end++) {
            Character cur = s.charAt(end);
            while (unique.contains(cur)) {
                unique.remove(s.charAt(start));
                start++;
            }
            unique.add(cur);
            size = Math.max(size, end - start + 1);
        }

        return size;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        
        i: int = 0
        j: int = 0
        max_window_size: int = 0
        window_chars: Set[str] = set()

        while j < len(s):
            if s[j] not in window_chars:
                window_chars.add(s[j])
                j += 1
            else:
                window_chars.remove(s[i])
                i += 1
            max_window_size = max(max_window_size, j - i)

        return max_window_size
        